### PR TITLE
Functor analysis

### PR DESCRIPTION
**Pull request by darth-tytus**
- Push some of the state from functors' instances into an analysis `FunctorAnalysis`. 

The long term idea is to replace the `PolymorphicObjectsTransformer` with some analyses and thus make `ast` purer.

Built on top of @b-scholz branch `b-scholz/FunctorAnalysis`.

### CODE REVIEW

File: src/ast/analysis/Functor.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+#pragma once
+
+#include "ast/IntrinsicFunctor.h"
+#include "ast/Program.h"
+#include "ast/UserDefinedFunctor.h"
+#include "ast/analysis/Analysis.h"
+#include "ast/utility/Visitor.h"
+#include <memory>
+#include <sstream>
```
Comment (mmcgr) [inline src/ast/analysis/Functor.h:25]: This isn't used. Did you mean `ostream`?

File: src/ast/analysis/Functor.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2020 The Souffle Developers. All rights reserved
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file Functor.h
```
Comment (mmcgr) [inline src/ast/analysis/Functor.cpp:11]:
```suggestion
 * @file Functor.cpp
```

File: src/ast/analysis/Functor.h (hunk 1 above)
Comment (darth-tytus) [reply src/ast/analysis/Functor.h]: Yes ^^

File: src/ast/IntrinsicFunctor.h [hunk 3]
```diff
@@ -61,37 +62,23 @@ class IntrinsicFunctor : public Functor {
...
 
 protected:
-    IntrinsicFunctor(
-            std::string op, const IntrinsicFunctorInfo* info, VecOwn<Argument> args, SrcLocation loc = {})
-            : Functor(std::move(args), std::move(loc)), function(std::move(op)), info(info) {
-        assert((!info || info->symbol == function) && "functor info must match symbol");
-    }
+    IntrinsicFunctor(std::string function, const std::optional<FunctorOp> op, VecOwn<Argument> args,
+            SrcLocation loc = {})
+            : Functor(std::move(args), std::move(loc)), function(std::move(function)), op(std::move(op)) {}
```
Comment (mmcgr) [inline src/ast/IntrinsicFunctor.h:81]:
```suggestion
    IntrinsicFunctor(std::string function, std::optional<FunctorOp> op, VecOwn<Argument> args,
            SrcLocation loc = {})
            : Functor(std::move(args), std::move(loc)), function(std::move(function)), op(op) {}
```
op is const so the move has no effect. It's also trivially copyable, which would also stop std::move having an effect

**Review by mmcgr — APPROVED** [review]

**Pull request merged**